src/x3d/x3dloadinternalstarling.pas

```diff
@@ -0,0 +1,611 @@
+      CoordInterp.FdKey.Items.Add(Key);
+      TexCoordInterp.FdKey.Items.Add(Key);
+    end;
+  end;
+
+  { This way, we have keys like
+    0 0.333 0.333 0.666 0.666 1
+    That is, all keys are repeated, except 0 and 1. }
+  CoordInterp.FdKey.Items.Add(1.0);
+  TexCoordInterp.FdKey.Items.Add(1.0);
+
+  { Add TimeSensor, CoordinateInterpolatorNode,
+    CoordinateInterpolator2DNode to Root node }
+  FRoot.AddChildren(TimeSensor);
+  FRoot.AddChildren(CoordInterp);
+  FRoot.AddChildren(TexCoordInterp);
+  AddRoutes(TimeSensor, CoordInterp, TexCoordInterp);
+end;
+
+procedure TStarlingTextureAtlasLoader.AddRoutes(
+  const TimeSensor: TTimeSensorNode;
+  const CoordInterp: TCoordinateInterpolatorNode;
+  const TexCoordInterp: TCoordinateInterpolator2DNode);
+var
+  R1, R2, R3, R4: TX3DRoute;
+begin
+  { Create routes. }
+  R1 := TX3DRoute.Create;
+  R2 := TX3DRoute.Create;
+  R3 := TX3DRoute.Create;
+  R4 := TX3DRoute.Create;
+  R1.SetSourceDirectly(TimeSensor.EventFraction_changed);
+  R1.SetDestinationDirectly(CoordInterp.EventSet_fraction);
+  R2.SetSourceDirectly(TimeSensor.EventFraction_changed);
+  R2.SetDestinationDirectly(TexCoordInterp.EventSet_fraction);
+  R3.SetSourceDirectly(CoordInterp.EventValue_changed);
+  R3.SetDestinationDirectly(FShapeCoord.FdPoint);
+  R4.SetSourceDirectly(TexCoordInterp.EventValue_changed);
+  R4.SetDestinationDirectly(FShapeTexCoord.FdPoint);
+  FRoot.AddRoute(R1);
+  FRoot.AddRoute(R2);
+  FRoot.AddRoute(R3);
+  FRoot.AddRoute(R4);
+end;
+
+function TStarlingTextureAtlasLoader.CheckAnimationNameAvailable(
+    const AnimationName: String): Boolean;
+begin
+  if FAnimationList.IndexOf(AnimationName) > -1 then
+  begin
+    WritelnWarning('Starling', 'Mixed animations tags (animation: %s) in "%s".', [AnimationName, FDisplayURL]);
+    Exit(false);
+  end;
+
+  FAnimationList.Add(AnimationName);
+  Result := true;
+end;
+
+
+constructor TStarlingTextureAtlasLoader.Create(const URL: String);
+begin
+  FURL := URL;
+  FDisplayURL := URIDisplay(FURL);
+
+  FSubTexture := TStarlingSubTexture.Create;
+  SetLength(FCoordArray, 6);
+  SetLength(FTexCoordArray, 6);
+  FAnimationList := TStringList.Create;
+end;
+
+function TStarlingTextureAtlasLoader.Load: TX3DRootNode;
+var
+  Doc: TXMLDocument;
+  Node: TDOMNode;
+  AtlasNode: TDOMElement;
+  I: TXMLElementIterator;
+  LastAnimationName: String;
+  CurrentAnimFrameCount: Integer;
+  TimeSensor: TTimeSensorNode;
+  CoordInterp: TCoordinateInterpolatorNode;
+  TexCoordInterp: TCoordinateInterpolator2DNode;
+  FirstFrameInFirstAnimation: Boolean;
+begin
+  ReadImportSettings;
+
+  FRoot := nil;
+  Doc := nil;
+  try
+    try
+      FRoot := TX3DRootNode.Create;
+      Doc := URLReadXML(FURL);
+      Node := Doc.FindNode('TextureAtlas');
+      if (Node = nil) or (Node.NodeType <> ELEMENT_NODE) then
+        raise EInvalidStarlingXml.CreateFmt('Invalid Starling XML file "%s" - TextureAtlas node not found.', [FDisplayURL]);
+      AtlasNode := Node as TDOMElement;
```

Comment:
If this is root, then just `AtlasNode := Doc.DocumentElement;`  without any typecast should be OK.